internal_workflow_client.go

```diff
@@ -245,6 +250,65 @@ GetHistoryLoop:
 	return history, nil
 }
 
+func (wc *workflowClient) GetWorkflowStackTrace(workflowID string, runID string, atDecisionTaskCompletedEventID int64) (string, error) {
+	getHistoryPage := newGetHistoryPageFunc(
+		wc.workflowService,
+		wc.domain,
+		&s.WorkflowExecution{WorkflowId: common.StringPtr(workflowID), RunId: common.StringPtr(runID)},
+		atDecisionTaskCompletedEventID,
+	)
+	return getWorkflowStackTraceImpl(workflowID, runID, getHistoryPage)
+}
+
+func getWorkflowStackTraceImpl(workflowID string, runID string, getHistoryPage GetHistoryPage) (string, error) {
+	firstPage, taskToken, err := getHistoryPage([]byte{})
+	if err != nil {
+		return "", err
+	}
+	if len(firstPage.Events) == 0 {
+		return "", errors.New("empty history")
+	}
+	startWorkflowEvent := firstPage.Events[0].WorkflowExecutionStartedEventAttributes
+	if startWorkflowEvent == nil {
+		return "", errors.New("First event is not WorkflowExecutionStarted")
+	}
+	registeredWorkflowFactory := newRegisteredWorkflowFactory()
+	workflowDefinitionFactory := getWorkflowDefinitionFactory(registeredWorkflowFactory)
+	logger, err := zap.NewDevelopment()
+	if err != nil {
+		return "", err
+	}
+	workerParams := workerExecutionParameters{
+		TaskList:                  startWorkflowEvent.GetTaskList().GetName(),
+		ConcurrentPollRoutineSize: defaultConcurrentPollRoutineSize,
+		Identity:                  startWorkflowEvent.GetIdentity(),
+		MetricsScope:              tally.NoopScope,
+		Logger:                    logger,
+		EnableLoggingInReplay:     false,
+		UserContext:               context.Background(),
+	}
+	var maxInt64 int64 = math.MaxInt64
+	taskHandler := newWorkflowTaskHandler(workflowDefinitionFactory, "unknown", workerParams, nil)
+	task := &s.PollForDecisionTaskResponse{
+		History:                firstPage,
+		PreviousStartedEventId: &maxInt64,
+		StartedEventId:         &maxInt64,
+		WorkflowExecution:      &s.WorkflowExecution{WorkflowId: &workflowID, RunId: &runID},
+		WorkflowType:           startWorkflowEvent.WorkflowType,
+		NextPageToken:          taskToken,
+	}
+	response, stackTrace, err := taskHandler.ProcessWorkflowTask(task, getHistoryPage, true)
+	if err == nil && response != nil && len(response.GetDecisions()) > 0 && len(stackTrace) == 0 {
+		lastDecision := response.GetDecisions()[0]
+		if lastDecision.GetDecisionType() == s.DecisionType_FailWorkflowExecution {
```

Comment:
why do we special handle this fail execution decision? Do we need to special handle all workflow termination decisions the same way?